test/dispatch.jl

```diff
@@ -0,0 +1,156 @@
+@testset "type_morespecific" begin
+    @testset "subtype" begin
+        @test type_morespecific(Int, Integer)
+        @test !type_morespecific(Integer, Int)
+        @test !type_morespecific(Int, Int)
+    end
+
+    @testset "number of arguments" begin
+        @test type_morespecific(Tuple{Int,AbstractFloat}, Tuple{Integer})
+    end
+
+    @testset "ambiguous" begin
+        @test !type_morespecific(Tuple{Int,AbstractFloat}, Tuple{Integer,Float64})
+        @test !type_morespecific(Tuple{Integer,Float64}, Tuple{Int,AbstractFloat})
+    end
+
+    @testset "length" begin
+        @test type_morespecific(Tuple{Int,Int}, Tuple{Int,Vararg{Int}})
+    end
+
+    @testset "union" begin
+        @test type_morespecific(Tuple{Int}, Tuple{Union{Int,Float64}})
+        @test type_morespecific(Tuple{Union{Int,Float64}}, Tuple{Integer})
+    end
+
+    @testset "union all" begin
+        @test type_morespecific(Tuple{Int}, Tuple{T} where T <: Integer)
+
+        @test !type_morespecific(Tuple{Integer}, Tuple{T} where T <: Integer)
+        @test !type_morespecific(Tuple{T} where T <: Integer, Tuple{Integer})
+
+        @test type_morespecific(Tuple{Type{Integer}}, Tuple{Type{T}} where T <: Integer)
+        @test !type_morespecific(Tuple{Type{T}} where T <: Integer, Tuple{Type{Integer}})
+    end
+end
+
+@testset "anonymous_signature" begin
+    @testset "diagonal dispatch" begin
+        diag(::T, ::T) where T <: Integer = nothing
+
+        diag_method = first(methods(diag, (Int, Int)))
+        diag_sig = Tuple{typeof(diag),T,T} where T <: Integer
+        @test diag_method.sig == diag_sig
+
+        @test anonymous_signature(Tuple{typeof(diag),Int,Int}) == Tuple{Int,Int}
+        @test anonymous_signature(diag_sig) == Tuple{T,T} where T <: Integer
+        @test anonymous_signature(diag_method) == Tuple{T,T} where T <:Integer
+    end
+
+    @testset "triangular dispatch" begin
+        tri(::Vector{T}, ::S) where {T, S<:T} = nothing
+
+        tri_method = first(methods(tri, (Vector, Int)))
+        tri_sig = Tuple{typeof(tri),Vector{T},S} where {T, S<:T}
+        @test tri_method.sig == tri_sig
+
+        @test anonymous_signature(Tuple{typeof(tri),Vector,Int}) == Tuple{Vector,Int}
+        @test anonymous_signature(tri_sig) == Tuple{Vector{T},S} where {T, S<:T}
+        @test anonymous_signature(tri_method) == Tuple{Vector{T},S} where {T, S<:T}
+    end
+end
+
+@testset "anon_morespecific" begin
+    plus_method = first(methods(+, (Int, Int)))
+    max_method = first(methods(max, (Int, Int)))
+
+    @test !type_morespecific(plus_method.sig, max_method.sig)
+    @test anon_morespecific(plus_method, max_method)
+end
+
+@testset "dispatch" begin
```

Comment:
I can't see any tests where the method is not the only method of the function.
The following is what i think is needed.

```
    @testset "more specific, method" begin
        f1(::Number) = 11
        f1(::Int) = 12
        f2(::Any) = 21
        f2(::Integer) = 22

        funcs = [f1, f2]

        m, f = dispatch(funcs, zero(Int))
        @test f == f1
        @test m == first(methods(f1, Tuple{Int}))

        m, f = dispatch(funcs, zero(UInt))
        @test f == f2
        @test m == first(methods(f2, Tuple{Integer}))
    end
```

An alternative to test the method idenditity might be to do 

```
@test Base.uncompressed_ast(m).code == [:(return 22)]
```

if one wanted to avoid having to trust `methods` with type tuples.
(Which one might, since it is a bit like testing the code on itself)

Reply:
This is being tested in "optional" but I can see what that isn't obvious